src/formats/CML.cpp

```diff
@@ -0,0 +1,581 @@
+    for (auto attribute : current.attributes()) {
+        auto name = std::string(attribute.name());
+        if (name == "id") { // do nothing, required junk for validation
+        } else if (name == "xmlns") {
+        } else if (name == "xmlns:cml") {
+        } else if (name == "xmlns:units") {
+        } else if (name == "xmlns:xsd") {
+        } else if (name == "xmlns:iupac") {
+        } else if (name == "title") {
+            frame.set("title", attribute.as_string());
+        } else {
+            warning("[CML] Unknown molecule attribute: " + name);
+        }
+    }
+
+    auto name_node = current.child("name");
+    if (name_node) {
+        frame.set("name", name_node.text().as_string());
+    }
+
+    auto crystal_node = current.child("crystal");
+    if (crystal_node) {
+        double a, b, c, alpha, beta, gamma;
+        a = b = c = alpha = beta = gamma = 0.0;
+        for (auto cell_param : crystal_node.children("scalar")) {
+            auto title = cell_param.attribute("title");
+            if (title) {
+                auto name = std::string(title.value());
+                if (name == "a") {
+                    a = cell_param.text().as_double();
+                } else if (name == "b") {
+                    b = cell_param.text().as_double();
+                } else if (name == "c") {
+                    c = cell_param.text().as_double();
+                } else if (name == "alpha") {
+                    alpha = cell_param.text().as_double();
+                } else if (name == "beta") {
+                    beta = cell_param.text().as_double();
+                } else if (name == "gamma") {
+                    gamma = cell_param.text().as_double();
+                } else {
+                    warning("[CML] Unknown crytal scalar: " + name);
+                }
+            }
+        }
+        frame.set_cell(UnitCell(a, b, c, alpha, beta, gamma));
+    }
+
+    // Frame level properties
+    auto properties_node = current.child("propertyList");
+    if (properties_node) {
+        for (auto property : properties_node.children("property")) {
+            auto title = property.attribute("title");
+            if (!title) {
+                warning("[CML] Skipping untitled property");
+                continue;
+            }
+
+            auto title_str = title.as_string();
+            auto scalar_node = property.child("scalar");
+            if (!scalar_node) {
+                warning("[CML] {} has no scalar associated with it!", title_str);
+                continue;
+            }
+
+            read_property_(frame, scalar_node, title_str);
+        }
+    }
+
+    // Read the atoms
+    std::unordered_map<std::string, size_t> ref_to_id;
+
+    // TODO: Read string vector versions instead of just the nodes
+    for (auto atom : current.child("atomArray").children("atom")) {
+        double x2, y2, x3, y3, z3, xf, yf, zf;
+        x2 = y2 = x3 = y3 = z3 = xf = yf = zf = 0.0;
```

Comment:
I would rather do this as `double x = 0, y = 0, z =0; bool is_frac = false`, and then set only x/y for x2,y2, set all three for x3/y3/z3, and set is_frac for xf/yf/zf

Reply:
I think I'm going to leave this as is for now given that we don't know what order we could get x2/x3/xf.

Reply:
Sounds good. I did not knew the all of them could be present in a file